source/dhtslib/memory.d

```diff
@@ -0,0 +1,86 @@
+module dhtslib.memory;
+
+import std.meta : AliasSeq, staticIndexOf;
+import std.traits : isPointer, isSomeFunction, ReturnType;
+import core.stdc.stdlib;
+import htslib;
+
+struct HtslibMemory(T, alias destroy)
+if(!isPointer!T && isSomeFunction!destroy)
+{
+    /// data pointer
+    T * ptr;
+
+    /// reference count pointer
+    private size_t * refct;
+
+    /// ability to use this as the ptr directly
+    alias ptr this;
+
+    /// ptr ctor
+    this(T * ptr)
+    {
+        this.ptr = ptr;
+        this.refct = cast(size_t*)malloc(size_t.sizeof);
+        *this.refct = 1;
+    }
+
+    /// struct ctor
+    this(T data)
+    {
+        *this.ptr = data;
```

Comment:
this will crash as `this.ptr is null`

Reply:
You are right. I need more unittests. 
> > I would highly recommend doing so. Getting your own destroy method is fairly simple as creating a struct that wraps a e.g. bam1_t and in its destructor calls bam_destroy

I mainly chose against when making this as I couldn't see how to best use it and maintain flexibility. Though now I see I could turn `HtslibMemory` into a template the achieves the same result and uses `RefCounted`.

Reply:
I have moved to using `RefCounted` for reference counting in 5e1a3ea. `HtslibMemory` is still a struct instead of a template as I previously considered above. Took me a minute of finagling to figure out how to still use the `HtslibMemory` type as the underlying `T *` type so I actually used the `RefCounted` implementation as a guide. I didn't realize that `alias this` could use a function.